File: examples/laser_tag/src/ajan_pomdp_planner/cpp/ajan_upperbound.cpp
//
// Created by bhuvanesh on 01.08.23.
// JNI Interface for DESPOT Tree Upperbound.
// See: examples/laser_tag/include/unified_pomdp_planner/tag_manhattan_upperbound.cpp
//

#include "ajan_upperbound.h"


namespace despot {
/* ==============================================================================
* AjanUpperBound class
* ==============================================================================*/
    AjanUpperBound::AjanUpperBound(const AjanAgent *model):
    tag_model_(model) {
        // TODO: Implement AjanUpperBound constructor to call JNI
        /**
         * For all States, PreCompute the distance between the rob and opp in value_[state_id]
        **/

    }

    double AjanUpperBound::Value(const despot::State &s) const {
        // Return precomputed value_[state.state_id];
        jobject javaState = AjanHelper::toJavaState(s);
        return AjanHelper::getEnv()->CallDoubleMethod(javaReferenceObject,
                                                      AjanHelper::getEnv()->GetMethodID(
                                                              AjanHelper::getUpperBoundClass(),
                                                              Value_.c_str(),
                                                              Value_State_Double_Sig.c_str()),
                                                              javaState);
    }

    double AjanUpperBound::Value(const despot::Belief *belief) const {
        double value = 0;
        /**
         * Loop through the particles and for each particle (aka. state) sum up the value
            by multiplying current value with pre_computed value_[state.state_id]. Refer paper Guess:Eq.15.
        **/
        jobject javaBelief = AjanHelper::toJavaAjanBelief(belief, nullptr);
        return AjanHelper::getEnv()->CallDoubleMethod(javaReferenceObject,AjanHelper::getMethodID(AJAN_UPPER_BOUND,Value_),javaBelief);
    }
}

File: examples/laser_tag/src/ajan_pomdp_planner/java/src/main/java/de/dfki/asr/ajan/pluginsystem/mdpplugin/utils/POMDP/DESPOT/implementation/AjanUpperBound.java
package de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.implementation;

import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.core.History;
import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.interface_.BeliefUpperBound;
import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.interface_.ParticleUpperBound;
import de.dfki.asr.ajan.pluginsystem.mdpplugin.utils.POMDP.DESPOT.interface_.State;
import jdk.jshell.spi.ExecutionControl;

import java.util.Vector;

public class AjanUpperBound extends ParticleUpperBound implements BeliefUpperBound {

    public long pointerToCpp;
    public AjanUpperBound(long cppReference){
        pointerToCpp = cppReference;
    }

    public AjanUpperBound() {
    }


    @Override
    protected double Value(State s) {
        // TODO: Implement AjanParticleUpperBound::Value Function for Particle to call JNI.
        // Return precomputed value_[state.state_id];
        System.err.println("AjanUpperBound.Value(State) is not implemented");
        return 0.0;
    }

    @Override
    public double Value(AjanBelief belief) {
        // TODO: Implement AjanParticleUpperBound::Value Function for Belief to call JNI.
        /*
         * Loop through the particles and for each particle (aka. state) sum up the value
         by multiplying current value with pre_computed value_[state.state_id]. Refer paper Guess:Eq.15.
         **/
        System.err.println("AjanUpperBound.Value(AjanBelief) is not implemented");
        return 0;
    }

    @Override
    public double Value(Vector<State> particles, History history) throws ExecutionControl.NotImplementedException {
        System.err.println("AjanUpperBound.Value(particles,history) is not implemented");
        return 0;
    }

    public void SetReferenceToCpp_(long ptr) {
        this.pointerToCpp = ptr;
    }
}
